feat!: Secure mode by default

kevin-dp opened the pull request:
Fixes [secure by default issue](https://github.com/electric-sql/electric/issues/2362).

Accompanying docs PR: https://github.com/electric-sql/electric/pull/2405

Introduces the `ELECTRIC_SECRET` env var that should be used to set an API secret that needs to be provided with all shape requests, unless we are running in insecure mode which can be set with an env var `ELECTRIC_INSECURE=true`.

**This is a breaking change because:**
- Electric must be started with one of the two env vars
- in secure mode, API calls must provide the secret on all shape requests

robacourt reviewed (approved):
Nicely done. Nice clear error/warning messages too 👍

KyleAMathews commented on integration-tests/tests/secure-mode.lux:
```diff
@@ -0,0 +1,62 @@
+  ??HTTP/1.1 401 Unauthorized
+
+  # rejected is secret is wrong
+  !curl -v -X GET "http://localhost:3000/v1/shape?table=items&offset=-1" --header "electric-secret: wrong-secret"
```
Let's make `electric-secret` a query parameter to keep it in line with cloud's `source_secret`.

KyleAMathews replied:
It's to avoid preflight requests which slow things down a lot https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request

It's not a big deal tbh either way as your client code isn't the one generally directly making the request to the electric server so the secret isn't shared.

I did want to make the change as in the future, there'll be a way for people to generate tokens specific to individual shapes e.g. for a non-logged in site where a lot of people are subscribed to the same data. We want to avoid the preflights in that scenario.

KyleAMathews reviewed (changes requested):
Thanks for picking this up! Looks great

kevin-dp commented:
TODO: generate a changeset

KyleAMathews commented on packages/sync-service/lib/electric/config.ex:
```diff
@@ -374,4 +374,23 @@ defmodule Electric.Config do
+        raise "Cannot set both ELECTRIC_SECRET and ELECTRIC_INSECURE=true"
+
+      !insecure && secret == nil ->
+        raise "Must set ELECTRIC_SECRET unless ELECTRIC_INSECURE=true. Setting ELECTRIC_INSECURE=true risks exposing your database, only use insecure mode if you know what you are doing."
```
```suggestion
        raise "You must set ELECTRIC_SECRET unless ELECTRIC_INSECURE=true. Setting ELECTRIC_INSECURE=true risks exposing your database, only use insecure mode in development or you've otherwise secured the Electric API"
```

KyleAMathews commented on packages/sync-service/lib/electric/config.ex:
```diff
@@ -374,4 +374,23 @@ defmodule Electric.Config do
+      true ->
+        if insecure do
+          Logger.warning(
+            "Running in insecure mode, this risks exposing your database, only use insecure mode if you know what you are doing."
```
```suggestion
            "Electric is running in insecure mode — this risks exposing your database — only use insecure mode in development or if you've otherwise secured the Electric API."
```

KyleAMathews commented on packages/sync-service/lib/electric/config.ex:
```diff
@@ -374,4 +374,23 @@ defmodule Electric.Config do
+  def validate_security_config!(secret, insecure) do
+    cond do
+      insecure && secret != nil ->
+        raise "Cannot set both ELECTRIC_SECRET and ELECTRIC_INSECURE=true"
```
```suggestion
        raise "You cannot set both ELECTRIC_SECRET and ELECTRIC_INSECURE=true"
```

KyleAMathews reviewed (commented):
Some copy suggestions

kevin-dp commented:
@KyleAMathews I moved the secret to a query parameter named `api_secret`. I used an underscore to be consistent with naming of `source_secret` in cloud. Let me know if you would prefer `api-secret` or another name.

kevin-dp commented:
In this PR i also added a `secret` option to the `ShapeStream`.
Not sure if that's a good idea because this secret should generally not be sent from the client code but from a secure proxy. Having it on the `ShapeStream` client may lead people to believe it's fine to use it. Shall i remove it?

KyleAMathews commented:
> @KyleAMathews I moved the secret to a query parameter named api_secret. I used an underscore to be consistent with naming of source_secret in cloud. Let me know if you would prefer api-secret or another name.

`api_secret` is great 👍

KyleAMathews commented:
> In this PR i also added a secret option to the ShapeStream.
Not sure if that's a good idea because this secret should generally not be sent from the client code but from a secure proxy. Having it on the ShapeStream client may lead people to believe it's fine to use it. Shall i remove it?

Yes :-D we've stripped out all special options in ShapeStream after we added the `headers` & `params` options — people can just add `api_secret` to the params object if they're querying the server directly.

KyleAMathews commented:
Can you post a screenshot of the logs when it boots up in insecure and secure mode? Since it's pretty critical people see this, we might want to put a box around the text.

kevin-dp commented:
> > In this PR i also added a secret option to the ShapeStream.
> > Not sure if that's a good idea because this secret should generally not be sent from the client code but from a secure proxy. Having it on the ShapeStream client may lead people to believe it's fine to use it. Shall i remove it?
> 
> Yes :-D we've stripped out all special options in ShapeStream after we added the `headers` & `params` options — people can just add `api_secret` to the params object if they're querying the server directly.

Ok, removing it!

kevin-dp commented:
> Can you post a screenshot of the logs when it boots up in insecure and secure mode? Since it's pretty critical people see this, we might want to put a box around the text.

Secure mode:
![Screenshot 2025-03-06 at 15 30 51](https://github.com/user-attachments/assets/480bcd38-218e-4709-98d6-c18104030795)

Insecure mode:
![Screenshot 2025-03-06 at 15 31 23](https://github.com/user-attachments/assets/ca4904b8-9a00-4e2a-a9b8-8fa9cf575f97)

KyleAMathews commented:
Hmm yeah that's probably fine actually 👍

KyleAMathews commented on packages/sync-service/lib/electric/config.ex:
```diff
@@ -374,4 +374,23 @@ defmodule Electric.Config do
+        raise "You cannot set both ELECTRIC_SECRET and ELECTRIC_INSECURE=true"
+
+      !insecure && secret == nil ->
+        raise "You must set ELECTRIC_SECRET unless ELECTRIC_INSECURE=true. Setting ELECTRIC_INSECURE=true risks exposing your database, only use insecure mode in development or you've otherwise secured the Electric API"
```
Another small copy tweak:

```suggestion
        raise "You must set ELECTRIC_SECRET or set ELECTRIC_INSECURE=true. Setting ELECTRIC_INSECURE=true risks exposing your database, only use insecure mode in development or you've otherwise secured the Electric API"
```

balegas commented:
Let's make this optional now and only enforce it for v1.

It would be nice to also print a warning message like the above saying that the default behavior is going to change for v1.
